### PR TITLE
add a one-shot database setup script

### PR DESCRIPTION
**Pull request by linki**
a ruby script living at `bin/setup_database` that set's up your database, surprise.

tldr; setup your database con and ruby environment, run the script and your done
# what it does
- creates the database if needed (in your env (note: it also creates test db when running in dev))
- ensures `schema.rb` exists (see below)
- for new databases: creates all tables and adds seed data (e.g. when setting up a new openproject instance)
- for existing databases: runs pending migrations if there are any (e.g. after new deployment)
# preconditions
- ensure postgres/mysql server is running
- configure db connection (`database.yml`, env vars)
- bundle

then run `./bin/setup_database`
# notes

script might look weird at first glance but it's actually very handy.

first of all it differentiates between new databases and existing databases. new databases need to run seeds/demo data. existing ones don't. both need to create all tables, of course. people will say that seed data is created on-demand (recreated if not already there) but in practice it doesn't really work. think about demo data as seeds. when i delete the demo project i don't want it to come back again. so at the very least you need to differentiate between seeds that have to be there (e.g. admin account) and demo data (default categories and such...)

anyways, second major point is our missing `schema.rb`. the second part of the script assumes we have a working `schema.rb`/`db/seeds` combination so we can run `rake db:setup`. but we don't. so on fresh OP checkouts we need to create `schema.rb` first (one way of doing that is by migrating the db, but that's just coincidence..). long story short, i expect `schema.rb` to be checked in at some poitn and then this script can become simpler. for now, it just does what it should in most if not all cases.
# caveats

adding seed data into `db/seeds` won't be executed for existing instances <= let's discuss this

this is just the first shot, need to flesh out the details, happy to receive feedback. more doc in the file.
# goal

the goal is to have a script that i can fire and forget whenever i want and it does the right thing. my target was to have an easy way to bootstrap openproject in a container. @oliverguenther pointed out another use case: switching between different OP versions and/or plugin combinations requires down- and up-migration between the different migration sets.

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://linki-feature-setup-database.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://linki-feature-setup-database.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

File: bin/setup_database [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  # or a database pending some migrations. in case of a new database we run
+  # db:setup, which will db:schema:load and then seed. for existing datbases
+  # we run migrations and don't seed (because we did that in the past)
+  if fresh_database
+    # system "#{rake} db:setup"
+
+    # so far for the theory, db:setup doesn't work right now because we have
+    # seed data in our migrations which would be missing. in order to move
+    # forward, we do it the old way and hope we can use the line above soon.
+    if pending_migrations
```
Comment (linki) [inline bin/setup_database:67]: this will probably always be `true` at this point

**Comment by kgalli** [comment]
:+1: I like  your approach. Maybe we can get this run db seeds or not question solved here once and for all and put the result into this PR.

**Comment by linki** [comment]
@kgalli i fear that this won't happen soon.

**Comment by NobodysNightmare** [comment]
I did not quite understand what @kgalli said, but the script looks like a nice helper :+1:

**Comment by oliverguenther** [comment]
If I understood that correctly, it can run seeds only on new instances, because there is no way to diff the seeds on file (when extended) with the state in the database. That would require seeds to take some form of migrations that can be applied as a delta update.

That script looks very promising and might be useful for what I intend to build: Automatically deploy development environments when switching branches (e.g., to avoid having to migrate and downgrate from plugins to core schema).

**Comment by basiszwo** [comment]
Closing this as we need to build bootstrap tasks anyway. We can come back to this PR as a source of inspiration.
